### ISSUE TITLE
No autocomplete for pandas column name in Jupyter notebook using bracket notation ['col']

### ISSUE THREAD

**Issue by natwille1** [microsoft/vscode-jupyter#8983]
## Environment data

-   VS Code version: 1.64.00
-   Jupyter Extension version (available under the Extensions sidebar):  v2022.1.1001821375
-   Python Extension version (available under the Extensions sidebar): v2022.0.1814523869
-   OS (Windows | Mac | Linux distro) and version: Windows 10
-   Python and/or Anaconda version: python 3.8.12
-   Type of virtual environment used (N/A | venv | virtualenv | conda | ...): conda
-   Jupyter server running: Local

## Expected behaviour

Autocomplete works for pandas dataframes in jupyter notebooks.

## Actual behaviour

Autocomplete does not work. Users have to manually type the column names every time. It even doesn't work when I've manually specified the column name several times. I've tried all the language servers, including Jedi, Pylance, and the default option. No autocomplete. 

## Steps to reproduce:

[**NOTE**: Self-contained, minimal reproducing code samples are **extremely** helpful and will expedite addressing your issue]

1. Create pandas dataframe 

`df = pd.DataFrame({'test': [1,2], 'test2': [2,3]})`

2. Call dataframe variable and check if autocomplete works for the columns

`df['test']`

![autocomplete_fail](https://user-images.githubusercontent.com/36617813/153467896-c47f08db-e4dd-4bb0-aa13-90d9b9a16b3c.gif)

## Logs

<details>

<summary>Output for <code>Jupyter</code> in the <code>Output</code> panel (<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Jupyter</code>)
</summary>

<p>

```
XXX
```

</p>
</details>

**Comment by IanMatthewHuff** [comment]
@natwille1 I repro the same. We have . completion after the column is name is completed but no autocompletion for the name itself when typing it as df['column_name']. We do have auto complete for just df.column_name, but that doesn't look like what you are looking for here:
<img width="590" alt="image" src="https://user-images.githubusercontent.com/812783/153470593-fb02ab43-4f21-4590-a6c6-fe2aeb1298d4.png">
Just to check do you get this ['column_name'] auto complete in a Jupyter notebook outside of vscode? I would assume this is the case, but in the Jupyter install that I have locally I don't see autocompletions for column names, might just be a misconfiguration that I have.

**Comment by IanMatthewHuff** [comment]
I have my language server set to pylance. Pylance should be using the python interpreter that you have selected for the current workspace so there could be an issue here if the kernel you have selected doesn't line up with the currently selected python interpreter. For the interactive inline execution were you looking for completions in the .py file itself? Or in the input box of the interactive window? If the completions of both types are missing in the .py file that would be an issue that we might have to transfer to the pylance team to investigate.

**Comment by natwille1** [comment]
Okay, my settings for python language server are "Default" which should use pylance given I have the extension installed. How can I check that the kernel and the selected python interpreter match? I use conda and select a specific environment for a specific project in the dropdown menu in the top right hand corner of the jupyter notebook. 

For interactive inline execution, I do not see any completions in the jupyter code cell in the .py file, but I do get the df.column_name completion in the input box of the interactive window. Thanks

**Comment by ido-greenfeld** [comment]
Jumping in here, I do have autocompletion of column names but it buried in other suggestions that are not relate to my dataframe.
This is what I get when I press ctrl+space inside the '' - 
![image](https://user-images.githubusercontent.com/53991912/153951358-664461ff-49b0-4dfe-aa17-15a87509bb93.png)

**Comment by natwille1** [comment]
Interesting - I never see this. What could be the factors affecting this behaviour? I can think of language server but I think we're also using pylance? Anything else?

**Comment by IanMatthewHuff** [comment]
@natwille1 Could you take a look at the Python Language Service section of the output window? When that starts up Pylance there is some logging about the interpreter that is being used, as well as any possible errors. Here is an example from my machine:

```
[Info  - 12:45:21 PM] (61504) Pylance language server 2022.2.2-pre.1 (pyright 618f54ed) starting
[Info  - 12:45:21 PM] (61504) Server root directory: /Users/ianhuff/.vscode-insiders/extensions/ms-python.vscode-pylance-2022.2.2-pre.1/dist
[Info  - 12:45:21 PM] (61504) No configuration file found.
[Info  - 12:45:21 PM] (61504) No pyproject.toml file found.
[Info  - 12:45:21 PM] (61504) Setting pythonPath for service "DebuggingDemo": "/Users/ianhuff/Documents/DataScience/DebuggingDemo/.debugEnv/bin/python"
[Warn  - 12:45:21 PM] (61504) stubPath /Users/ianhuff/Documents/DataScience/DebuggingDemo/typings is not a valid directory.
[Info  - 12:45:21 PM] (61504) Assuming Python version 3.7
[Info  - 12:45:21 PM] (61504) Assuming Python platform Darwin
```

**Comment by natwille1** [comment]
Yep mine looks the same as yours.. :/ 


[Info  - 09:53:47] (20496) Pylance language server 2022.2.3 (pyright 9b0ea9eb) starting
[Info  - 09:53:47] (20496) Server root directory: c:\Users\nwillems\.vscode\extensions\ms-python.vscode-pylance-2022.2.3\dist
[Info  - 09:53:47] (20496) No configuration file found.
[Info  - 09:53:47] (20496) No pyproject.toml file found.
[Info  - 09:53:47] (20496) Setting pythonPath for service "proms-prediction": "C:\Users\nwillems\Anaconda3\envs\databricks\python.exe"
[Warn  - 09:53:47] (20496) stubPath c:\Users\nwillems\dev\proms-prediction\typings is not a valid directory.
[Info  - 09:53:48] (20496) Assuming Python version 3.8
[Info  - 09:53:48] (20496) Assuming Python platform Windows

**Comment by rchiodo** [comment]
I can get completions to work in a jupyter notebook (outside of VS code) and it does support the ```df.['column_name']``` syntax:

![image](https://user-images.githubusercontent.com/19672699/154748968-feb42f05-e617-4cba-a44d-f0f178e63503.png)

(Jupyter brings up completions on the 'tab' key)

So it would seem like we could implement this.

**Comment by IanMatthewHuff** [comment]
Sounds good I wasn't seeing the [''] syntax in my Jupyter install but it was having a bunch of auto complete issues in general. I'm going to leave this issue open for supporting the bracket syntax, and I'm going to file a new issue to look at if we want to consider adding Jupyter completions into .py files when using the interactive window so we can track those issues separately.

**Comment by natwille1** [comment]
> @natwille1 My apologies. I think I need to back up a second to make sure that I'm looking at your specific scenario. Do you mind if I back up for a second and restate?
> 
> From your comments (all of this is referring to inside of VS Code):
> 
> * In a Jupyter notebook you don't see completions for something like df['column_name'] when trying to complete the column name.
> * In a Jupyter notebook you do see completions for df.column_name, but it's slow and take a few seconds to come up.
> * When executing cell blocks from a .py file in the Interactive Window you don't see any column completions for the df at all either with a "." or with ['. But you do see other completions on the df (like df. and you would see 'abs' and other functions on the df).
> * When executing cell blocks from a .py fine in the Interactive Window in the Input box at the bottom of the Interactive Window you see the same . completions as in the notebook file.
> 
> Is that the right summary? I've looking at these scenarios and I think that I repro basically the same as you do. I think that I understand why most of the differences are happening, but I agree that there is some confusion about why things are seen in some scenarios and not others.
> 
> * Right now we don't provide column name completions with the ['column_name'] syntax. So that's just a missing feature
> * In the notebook and the input box intellisense is being provided by both pylance and by the running jupyter server. Jupyter has their own completions and we combine those with the static analysis from pylance. That's why in those scenarios you see .column_name as an option. However in .py file intellisense has only pylance so you see the member functions but not the dynamically added function names. Their might be a way to change this where if an Interactive Window is running we provide the same Jupyter completions to the .py file, but we don't do that now.
> * The fact that completions are coming from Jupyter might be related to why they are slow to come back. For those Jupyter specific completions like the column names we are at the mercy of how fast the Jupyter server completion is, we can't control that directly.
> 
> @rchiodo

Correct description and thanks for outlining why and creating a new issue around this - looking forward to seeing any new support for these features!

**Comment by mazin-abdelghany** [comment]
In relation to this open issue, Jupyter Notebooks are also able to tab autocomplete files that are in the same working directory as the open notebook. 

For example, if the file `example.csv` is in the same working directory as the the open .ipynb file, `pd.read_csv('exam')` + `tab` will autocompletes the file name within the `pd.read_csv()` call to `example.csv`.

The Jupyter Notebook extension in VS Code does not replicate this behavior.

**Comment by DonJayamanne** [comment]
**Internal notes for the team to discuss**

I think we should revert the change that fixed https://github.com/microsoft/vscode-jupyter/issues/7136
That was merely a paper cut from an internal user, and not something that was reported by users, and as a result of that change we now have code completion not appearing when it should. I.e. this is a regression in functionality now, as a result of a fix for a paper cut. 

My suggestion is to first unblock the users) then look into a long term proper fix.

**Comment by rchiodo** [comment]
We could change this so that:
- Don't filter to just file paths in strings
- Make the default trigger for completions to always include string chars
```
"jupyter.pythonCompletionTriggerCharacters": ".%\"'"
```

**Comment by rchiodo** [comment]
Not super happy with this. Jupyter completions suck.

This is the result I get if I 'fix' this and put it back how it was:

![image](https://user-images.githubusercontent.com/19672699/170345665-acda9dcb-3aad-4f91-800e-795776e7db4a.png)

As you can see it's listing out every possible match for the letter 'a'. Not just columns. Jupyter isn't filtering anything.

Jupyter notebook is more precise though. It gives me this:

![image](https://user-images.githubusercontent.com/19672699/170355574-00930fab-7b2b-4d55-939e-f201a5ba512b.png)

And the difference is in our use of Jedi. We disable this because of perf problems with Jedi in a kernel. If I reenable it, our completions come out the same as Jupyter

![image](https://user-images.githubusercontent.com/19672699/170355763-6b9296f2-a692-4120-a1e5-a7473b8db4d9.png)

I'm going to add a setting to enable more precise intellisense. With a warning that this makes everything potentially slower.

**Comment by rchiodo** [comment]
Verification steps:

- Enable this setting:
```json
"jupyter.enableExtendedKernelCompletions": true
```
- Follow the steps above for creating a dataframe and trying to enter the columns

**Comment by tanhakabir** [comment]
I have the setting enabled but weirdly I'm getting suggestions for all the folders I have in my GitHub repo: 
<img width="927" alt="image" src="https://user-images.githubusercontent.com/12758612/171751162-8a8cb3d9-9163-434d-b07f-8229b88b3fac.png">

**Comment by rchiodo** [comment]
That's by design. Jupyter isn't smart enough to filter them out.

**Comment by rchiodo** [comment]
The verification would be that the column names do show up in that list. It looks like you had 'test' and 'test2'? They did show up?

**Comment by rebornix** [comment]
@rchiodo it works well but still reproduces with `python.pylanceLspNotebooksEnabled` enabled.

**Comment by rchiodo** [comment]
@rebornix can you explain what you mean by still reproduces? I have that setting enabled and it works for me:

![image](https://user-images.githubusercontent.com/19672699/171921790-670014cf-d8b5-497a-b33e-c5c235a5bef1.png)

**Comment by rebornix** [comment]
@rchiodo I was not seeing `test` and `test2` with the sample in https://github.com/microsoft/vscode-jupyter/issues/8983#issuecomment-1035292185

**Comment by rchiodo** [comment]
It works for me? Did you run the cell that generated the DF first? 

![image](https://user-images.githubusercontent.com/19672699/171926465-7f66530a-e5b5-45f1-97fd-fa279a7b3714.png)

These completions come from the kernel. You have to create the DF before it will know about the columns.

**Comment by StevenLi-DS** [comment]
This does not work, or at least is very unstable. Sometimes, it works in the editor but not the interactive window, and sometimes the other way around. Very often... Neither dot nor brackets notion works at all.
